src/extras/relayPagination.ts

```diff
@@ -0,0 +1,109 @@
+import { warning } from '../helpers/warning';
+import { Resolver, NullArray, Variables } from '../types';
+
+interface ConnectionArgs {
+  first?: number;
+  last?: number;
+  before?: string;
+  after?: string;
+}
+
+export const relayPagination = (): Resolver => {
+  return (_, args, cache, info) => {
+    const { parentKey: key, fieldName } = info;
+    const { first, last } = args as ConnectionArgs;
+
+    if (typeof first === 'number' && typeof last === 'number') {
+      warning(
+        false,
+        'relayPagination(...) was used with both a `first` and `last` argument,\n' +
+          'but it must not be used with both at the same time.'
+      );
+
+      return null;
+    } else if (typeof first !== 'number' && typeof last !== 'number') {
+      warning(
+        false,
+        'relayPagination(...) was used without a `first` and `last` argument,\n' +
+          'but it must not be used with at least one of them.'
+      );
+
+      return null;
+    }
+
+    const isForwardPagination = typeof first === 'number';
+    const size = (isForwardPagination ? first : last) || 0;
+
+    if (size <= 0) {
+      warning(
+        false,
+        'relayPagination(...) was used with a `first` or `last` argument that is less than zero.'
+      );
+    }
+
+    const childArgs: Variables = {};
+    if (typeof first === 'number') childArgs.first = first;
+    if (typeof last === 'number') childArgs.last = last;
+
+    const mergedEdges: NullArray<string> = [];
+
+    let dataKey = cache.resolve(key, fieldName, childArgs) as string;
+    let edgesKeys = cache.resolve(dataKey, 'edges') as string;
+    let infoKey = cache.resolve(dataKey, 'pageInfo') as string;
+    let prevInfoKey: string | null = null;
+
+    const connectionType = cache.resolve(dataKey, '__typename');
+    if (!connectionType) {
+      return undefined;
+    }
+
+    while (dataKey !== null && Array.isArray(edgesKeys)) {
+      // Store the last valid info key
+      prevInfoKey = infoKey;
+
+      // Add the receives nodes to the list of nodes
+      if (isForwardPagination) {
+        for (let i = 0; i < size; i++) mergedEdges.push(edgesKeys[i]);
+      } else {
+        for (let i = size - 1; i >= 0; i--) mergedEdges.unshift(edgesKeys[i]);
```

Comment:
NIT: spitballing out here, unshift is very slow compared to push can't we always push and then reverse at the end?